abci/vote_extensions.go

```diff
@@ -44,7 +65,37 @@ func NewVoteExtensionHandler(mp VoteExtensionMempool, txDecoder sdk.TxDecoder,
 // returns it in its vote extension.
 func (h *VoteExtensionHandler) ExtendVoteHandler() ExtendVoteHandler {
 	return func(ctx sdk.Context, req *RequestExtendVote) (*ResponseExtendVote, error) {
-		panic("implement me")
+		var voteExtension []byte
+
+		// Reset the cache if necessary
+		h.resetCache(ctx.BlockHeight())
+
+		// Iterate through auction bids until we find a valid one
+		auctionIterator := h.mempool.AuctionBidSelect(ctx)
+
+		for ; auctionIterator != nil; auctionIterator = auctionIterator.Next() {
+			bidTx := auctionIterator.Tx()
+
+			// Verify the bid tx can be encoded and included in vote extension
+			bidBz, err := h.txEncoder(bidTx)
+			if err != nil {
+				continue
+			}
+
+			hashBz := sha256.Sum256(bidBz)
+			hash := hex.EncodeToString(hashBz[:])
+
+			// Validate the auction transaction and cache result
+			if err := h.verifyAuctionTx(ctx, bidTx); err != nil {
+				h.cache[hash] = err
+			} else {
+				h.cache[hash] = nil
```

Comment:
We're need to check and set the cache in `VerifyVoteExtension`, not `ExtendVote`. The cache is so we don't verify the same vote ext twice.

Also note, CometBFT will never call `ExtendVote` and `VerifyVoteExtension` on the _same client_ :)

Reply:
Ah, i was under the assumption that it might call it both on the same client so caching on both sides makes sense. Will remove.